pmlc/dialect/pxa/transforms/fusion.cc

```diff
@@ -0,0 +1,425 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+using mlir::AffineLoadOp;
+using mlir::AffineParallelOp;
+using mlir::Block;
+using mlir::BlockArgument;
+using mlir::Operation;
+using mlir::StrideInfo;
+
+namespace {
+
+using WriteRead = std::pair<AffineReduceOp, AffineLoadOp>;
+using WriteWrite = std::pair<AffineReduceOp, AffineReduceOp>;
+
+struct FusionInfo {
+  // The parallel for ops
+  AffineParallelOp apA;
+  AffineParallelOp apB;
```

Comment:
It might be a little easier to read by expanding `apA` to `affineParallelA`

Reply:
Actually, could we place all the state for a particular `affine.parallel` in its own struct?

like:
```
struct FusionAffineParallelInfo {
  AffineParallelOp op;
  llvm::SmallVector<int64_t, 4> sizes;
  llvm::SmallVector<int64_t, 8> tile;
};
```

Reply:
Probably a good idea, I'll try to refactor to something like that when I add the actual tiling support, think I will leave it for today though if that's OK.